src/main/java/io/smartcat/ranger/core/CompositeValue.java

```diff
@@ -0,0 +1,51 @@
+package io.smartcat.ranger.core;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import edu.emory.mathcs.backport.java.util.Collections;
+
+/**
+ * Composite value containing its child values.
+ */
+public class CompositeValue implements Value<Map<String, Object>> {
+
+    private final Map<String, Value<?>> values;
+    private final Map<String, Object> evaluatedValues;
+    private final Map<String, Object> immutableEvaluatedValues;
+
+    /**
+     * Constructs composite value with no initial child values.
+     */
+    public CompositeValue() {
+        this(new HashMap<>());
+    }
+
+    /**
+     * Constructs composite value with specified initial child values.
+     *
+     * @param values Initial child values.
+     */
+    @SuppressWarnings("unchecked")
+    public CompositeValue(Map<String, Value<?>> values) {
+        this.values = new HashMap<>(values);
+        this.evaluatedValues = new HashMap<>();
+        this.immutableEvaluatedValues = Collections.unmodifiableMap(this.evaluatedValues);
+    }
+
+    /**
+     * Sets new or updates existing child value of this composite value.
+     *
+     * @param name Name of the child value.
+     * @param value New child value.
+     */
+    public void setValue(String name, Value<?> value) {
```

Comment:
Guess better name would be `addValue`? Also I would add `removeValue` to have symmetric values